runtime/common/src/transfer_filter.rs

```diff
@@ -231,18 +236,53 @@ where
 			return Ok(());
 		};
 
-		pallet_transfer_allowlist::pallet::Pallet::<T>::allowance(
-			who.clone(),
-			Location::Local(recv.clone()),
-			FilterCurrency::All,
-		)
-		.or_else(|_| {
+		amalgamate_allowance(
+			pallet_transfer_allowlist::pallet::Pallet::<T>::allowance(
+				who.clone(),
+				Location::Local(recv.clone()),
+				FilterCurrency::All,
+			),
 			pallet_transfer_allowlist::pallet::Pallet::<T>::allowance(
 				who.clone(),
 				Location::Local(recv.clone()),
 				FilterCurrency::Specific(CurrencyId::Native),
-			)
-		})
+			),
+		)
 		.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Custom(255)))
 	}
 }
+
+fn amalgamate_allowance(
+	first: Result<Option<Location>, DispatchError>,
+	second: Result<Option<Location>, DispatchError>,
+) -> DispatchResult {
+	match (first, second) {
+		// There is an allowance set for `Specific(id)`, but NOT for the given recv
+		// There is an allowance set for `All`, but NOT for the given recv
+		(Err(e), Err(_)) => Err(e),
+		// There is an allowance set for `Specific(id)`, but NOT for the given recv
+		// There is an allowance set for `All`, for the given recv
+		(Err(_), Ok(Some(_))) => Ok(()),
+		// There is an allowance set for `Specific(id)`, for the given recv
+		// There is an allowance set for `All`, but NOT for the given recv
+		(Ok(Some(_)), Err(_)) => Ok(()),
+		// There is NO allowance set for `Specific(id)`
+		// There is an allowance set for `All`, but NOT for the given recv
+		(Ok(None), Err(e)) => Err(e),
+		// There is an allowance set for `Specific(id)`, but NOT for the given recv
+		// There is NO allowance set for `All`
+		(Err(e), Ok(None)) => Err(e),
+		// There is an allowance set for `Specific(id)`, for the given recv
+		// There is an allowance set for `All`, for the given recv
+		(Ok(Some(_)), Ok(Some(_))) => Ok(()),
+		// There is NO allowance set for `Specific(id)`
+		// There is NO allowance set for `All`
+		(Ok(None), Ok(None)) => Ok(()),
+		// There is an allowance set for `Specific(id)`, for the given recv
+		// There is NO allowance set for `All`
+		(Ok(Some(_)), Ok(None)) => Ok(()),
+		// There is NO allowance set for `Specific(id)`
+		// There is an allowance set for `All`, for the given recv
+		(Ok(None), Ok(Some(_))) => Ok(()),
+	}
+}
```

Comment:
I think there is a common factor here. If there is no `Err` or there is some `Some`, then `Ok` else `Err`, but not sure if you prefer to leave it explicitly.

Reply:
Sure!